design/service-apis.md

```diff
@@ -0,0 +1,249 @@
+
+## High-Level Design
+
+Contour Operator will manage Contour using the `gatewayclasses.networking.x-k8s.io` and `gateways.networking.x-k8s.io`
+CRDs. The combination of GatewayClass and Gateway CRs will provide the user interface for managing an instance of
+Contour. Service APIs provides extension points to allow for custom resources to be linked at various layers of the API.
+This provides more granular customization at the appropriate places within the API structure. These extension points
+will be leveraged to expose Contour-specific configuration.
+
+## Detailed Design
+
+### GatewayClass
+
+GatewayClass is a cluster-scoped resource that defines a set of Gateways sharing a common behavior. A GatewayClass
+with `controller: projectcontour.io/contour-operator` will be managed by Contour Operator. There MUST be at least one
+GatewayClass defined in order to be able to have a functional Gateway. This is similar to IngressClass for Ingress and
+StorageClass for PersistentVolumes. The following is an example of a GatewayClass managed by Contour Operator:
+
+```yaml
+kind: GatewayClass
+apiVersion: networking.x-k8s.io/v1alpha1
+metadata:
+  name: example-gatewayclass
+spec:
+  controller: projectcontour.io/contour-operator
+  parametersRef:
+    name: example-contour-config
+    group: operator.projectcontour.io
+    kind: ContourConfig
+```
+
+The `controller` field is a domain/path string that indicates the controller responsible for managing Gateways of this
+class. The `projectcontour.io/contour-operator` string will be used to specify Contour Operator as the controller
+responsible for handling the GatewayClass. This field is not mutable and cannot be empty. Versioning can be accomplished
+by encoding the operator's version into the path. For example:
+
+```
+projectcontour.io/contour-operator/v1   // Use version 1
+projectcontour.io/contour-operator/v2   // Use version 2
+projectcontour.io/contour-operator      // Use the default version
+```
+
+The `parametersRef` field is a reference to an implementation-specific resource containing configuration parameters that
+are associated to the GatewayClass. This field will be used to expose Contour-specific configuration. If the field is
+unspecified, the operator will create an instance of Contour with default settings. Since `parametersRef` does not
+contain a namespace field, the referenced resource must exist in a known namespace. The operator's namespace will be
+used as the known namespace. If the referenced resource cannot be found, the operator will set the GatewayClass
+`InvalidParameters` status condition to true.
+
+### ContourConfig
+
+A new ContourConfig CR will be introduced to provide a structured object for representing the Contour configuration.
+This configuration-only resource will be consumed by a GatewayClass. ContourConfig is equivalent to spec of a Contour CR
+to provide consistency and supportability for managing Contour using the Contour CR or Service APIs. For example:
```

Comment:
Should `ContourConfig` be spec of `Contour` for consistency between the APIs or should we maintain an entirely separate API just to hold the Contour configuration for Service APIs? Note that when I say "configuration" I mean the e2e config of a COntour environment and not https://github.com/projectcontour/contour/blob/main/examples/contour/01-contour-config.yaml.

cc: @jpeach @stevesloka

Reply:
Could they merge and remove the current config map as configuration? I think long term that's probably a better approach.

Reply:
> Could they merge and remove the current config map as configuration? 

@stevesloka even without adding support for Service APIs, the intent is to expose `contour.yaml` (i.e. Contour ConfigMap) configuration to users through the `Contour` custom resource. We want to be selective on what configuration we expose. This is why no `contour.yaml` was exposed through the API for v1.11.0.

Reply:
It's more of a `Gateway` config vs a `ContourConfig`, maybe that name is what's confusing since now we have a contour configmap and a contour config CRD.

Reply:
@stevesloka `gatewayclass.spec.parametersRef` is a reference to a resource containing implementation-specific config parameters. I chose `ContourConfig` instead of a name such as `GatewayConfig` since the parameters are implementation-specific.